internal/dag/cache.go

```diff
@@ -84,24 +84,89 @@ func (kc *KubernetesCache) init() {
 	kc.extensions = make(map[types.NamespacedName]*contour_api_v1alpha1.ExtensionService)
 }
 
-// matchesIngressClass returns true if the given Kubernetes object
-// belongs to the Ingress class that this cache is using.
-func (kc *KubernetesCache) matchesIngressClass(obj metav1.Object) bool {
+// matchesIngressClass returns true if the given IngressClass
+// is the one this cache is using.
+func (kc *KubernetesCache) matchesIngressClass(obj *networking_v1.IngressClass) bool {
+	// If no ingress class name set, we allow an ingress class that is named
+	// with the default Contour accepted name.
+	if kc.IngressClassName == "" && obj.Name == "contour" {
+		return true
+	}
+	// Otherwise, the name of the ingress class must match what has been
+	// configured.
+	if obj.Name == kc.IngressClassName {
+		return true
+	}
+	return false
+}
+
+// ingressMatchesIngressClass returns true if the given Ingress object matches
+// the configured ingress class name via annotation or Spec.IngressClassName.
+// Annotation takes precedence over the Spec value.
+func (kc *KubernetesCache) ingressMatchesIngressClass(obj metav1.Object) bool {
+	annotationClass := annotation.IngressClass(obj)
+	var specClass string
+	switch obj := obj.(type) {
+	case *v1beta1.Ingress:
+		specClass = pointer.StringPtrDerefOr(obj.Spec.IngressClassName, "")
+	case *networking_v1.Ingress:
+		specClass = pointer.StringPtrDerefOr(obj.Spec.IngressClassName, "")
+	}
+
+	// If annotation is set or annotation is not set and spec is not set, we
+	// will try to match the annotation.
+	checkAnnotation := annotationClass != "" || specClass == ""
```

Comment:
I think I was trying to only do the annotation match check once for the multiple cases where it is relevant and I felt it collapsed some logic together. I think also I tried to fall through to return `false` and log the debug message in one place and that made this a bit more complicated. Another way to do this would be the following, though it would require some duplication and checking whether we had a match or not to log the debug message:

Doing things another way it might look like this:
- if annotation is non-empty, check if it matches
  - return result of annotation helper
- if spec field is non-empty check if it matches
  - return result of checking class
- return true if no class configured

Reply:
Yeah to my eye that's a little easier to read since each case is handled directly. I see what you mean about the logging. You could even move the impl of getting the spec ingress class to a helper if you wanted (since we have one for the annotation, it keeps things at the same level of abstraction) but not a big deal.

Reply:
>  move the impl of getting the spec ingress class to a helper if you wanted

Yeah gonna have to do that since the status updater bits will need it as well